pgmpy/models/DiscreteBayesianNetwork.py

```diff
@@ -1176,23 +1176,50 @@ def do(self, nodes, inplace=False):
         OutEdgeView([('asia', 'tub'), ('tub', 'either'), ('smoke', 'lung'), ('lung', 'either'),
                      ('bronc', 'dysp'), ('either', 'xray'), ('either', 'dysp')])
         """
-        if isinstance(nodes, (str, int)):
-            nodes = [nodes]
-        else:
-            nodes = list(nodes)
-
-        if not set(nodes).issubset(set(self.nodes())):
+        if not isinstance(interventions, dict):
             raise ValueError(
-                f"Nodes not found in the model: {set(nodes) - set(self.nodes)}"
+                "`do` expects a dict of the form {variable: state}, "
+                f"got {type(interventions)}"
             )
 
+        nodes = list(interventions.keys())
+        states = list(interventions.values())
+
+        missing_nodes = set(nodes) - set(self.nodes())
+        if missing_nodes:
+            raise ValueError(f"Intervention variable(s) not in model: {missing_nodes}")
+
+        for var, forced_state in interventions.items():
+            cpd = self.get_cpds(node=var)
+            if forced_state not in cpd.state_names[var]:
```

Comment:
Can we move this check to the beginning of the method? I think it makes the main part of the logic a bit cleaner.